// n8n-nodes-base/dist/nodes/Google/Drive/test/v2/node/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMockExecuteFunction = exports.driveNode = void 0;
exports.createTestStream = createTestStream;
const lodash_1 = require("lodash");
const n8n_core_1 = require("n8n-core");
const stream_1 = require("stream");
exports.driveNode = {
    id: '11',
    name: 'Google Drive node',
    typeVersion: 3,
    type: 'n8n-nodes-base.googleDrive',
    position: [42, 42],
    parameters: {},
};
const createMockExecuteFunction = (nodeParameters, node, continueOnFail = false) => {
    const fakeExecuteFunction = {
        getNodeParameter(parameterName, _itemIndex, fallbackValue, options) {
            const parameter = options?.extractValue ? `${parameterName}.value` : parameterName;
            return (0, lodash_1.get)(nodeParameters, parameter, fallbackValue);
        },
        getNode() {
            return node;
        },
        helpers: {
            constructExecutionMetaData: n8n_core_1.constructExecutionMetaData,
            returnJsonArray: () => [],
            prepareBinaryData: () => { },
            httpRequest: () => { },
        },
        continueOnFail: () => continueOnFail,
    };
    return fakeExecuteFunction;
};
exports.createMockExecuteFunction = createMockExecuteFunction;
function createTestStream(byteSize) {
    let bytesSent = 0;
    const CHUNK_SIZE = 64 * 1024;
    return new stream_1.Readable({
        read() {
            const remainingBytes = byteSize - bytesSent;
            if (remainingBytes <= 0) {
                this.push(null);
                return;
            }
            const chunkSize = Math.min(CHUNK_SIZE, remainingBytes);
            const chunk = Buffer.alloc(chunkSize, 'A');
            bytesSent += chunkSize;
            this.push(chunk);
        },
    });
}
