// @n8n/n8n-nodes-langchain/dist/nodes/tools/ToolVectorStore/ToolVectorStore.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolVectorStore = void 0;
const chains_1 = require("langchain/chains");
const tools_1 = require("langchain/tools");
const logWrapper_1 = require("../../../utils/logWrapper");
const sharedFields_1 = require("../../../utils/sharedFields");
class ToolVectorStore {
    constructor() {
        this.description = {
            displayName: 'Vector Store Question Answer Tool',
            name: 'toolVectorStore',
            icon: 'fa:database',
            iconColor: 'black',
            group: ['transform'],
            version: [1],
            description: 'Answer questions with a vector store',
            defaults: {
                name: 'Answer questions with a vector store',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Tools'],
                    Tools: ['Other Tools'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.toolvectorstore/',
                        },
                    ],
                },
            },
            inputs: [
                {
                    displayName: 'Vector Store',
                    maxConnections: 1,
                    type: "ai_vectorStore",
                    required: true,
                },
                {
                    displayName: 'Model',
                    maxConnections: 1,
                    type: "ai_languageModel",
                    required: true,
                },
            ],
            outputs: ["ai_tool"],
            outputNames: ['Tool'],
            properties: [
                (0, sharedFields_1.getConnectionHintNoticeField)(["ai_agent"]),
                {
                    displayName: 'Data Name',
                    name: 'name',
                    type: 'string',
                    default: '',
                    placeholder: 'e.g. users_info',
                    validateType: 'string-alphanumeric',
                    description: 'Name of the data in vector store. This will be used to fill this tool description: Useful for when you need to answer questions about [name]. Whenever you need information about [data description], you should ALWAYS use this. Input should be a fully formed question.',
                },
                {
                    displayName: 'Description of Data',
                    name: 'description',
                    type: 'string',
                    default: '',
                    placeholder: "[Describe your data here, e.g. a user's name, email, etc.]",
                    description: 'Describe the data in vector store. This will be used to fill this tool description: Useful for when you need to answer questions about [name]. Whenever you need information about [data description], you should ALWAYS use this. Input should be a fully formed question.',
                    typeOptions: {
                        rows: 3,
                    },
                },
                {
                    displayName: 'Limit',
                    name: 'topK',
                    type: 'number',
                    default: 4,
                    description: 'The maximum number of results to return',
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        const name = this.getNodeParameter('name', itemIndex);
        const toolDescription = this.getNodeParameter('description', itemIndex);
        const topK = this.getNodeParameter('topK', itemIndex, 4);
        const vectorStore = (await this.getInputConnectionData("ai_vectorStore", itemIndex));
        const llm = (await this.getInputConnectionData("ai_languageModel", 0));
        const description = tools_1.VectorStoreQATool.getDescription(name, toolDescription);
        const vectorStoreTool = new tools_1.VectorStoreQATool(name, description, {
            llm,
            vectorStore,
        });
        vectorStoreTool.chain = chains_1.VectorDBQAChain.fromLLM(llm, vectorStore, {
            k: topK,
        });
        return {
            response: (0, logWrapper_1.logWrapper)(vectorStoreTool, this),
        };
    }
}
exports.ToolVectorStore = ToolVectorStore;
