// n8n-nodes-base/dist/credentials/AzureStorageOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureStorageOAuth2Api = void 0;
class AzureStorageOAuth2Api {
    constructor() {
        this.name = 'azureStorageOAuth2Api';
        this.displayName = 'Azure Storage OAuth2 API';
        this.extends = ['microsoftOAuth2Api'];
        this.documentationUrl = 'azurestorage';
        this.properties = [
            {
                displayName: 'Account',
                name: 'account',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Base URL',
                name: 'baseUrl',
                type: 'hidden',
                default: '=https://{{ $self["account"] }}.blob.core.windows.net',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: 'https://storage.azure.com/.default',
            },
        ];
    }
}
exports.AzureStorageOAuth2Api = AzureStorageOAuth2Api;
