// n8n-nodes-base/dist/credentials/CalendlyOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendlyOAuth2Api = void 0;
class CalendlyOAuth2Api {
    constructor() {
        this.name = 'calendlyOAuth2Api';
        this.extends = ['oAuth2Api'];
        this.displayName = 'Calendly OAuth2 API';
        this.documentationUrl = 'calendly';
        this.icon = 'file:icons/Calendly.svg';
        this.properties = [
            {
                displayName: 'Grant Type',
                name: 'grantType',
                type: 'hidden',
                default: 'authorizationCode',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://auth.calendly.com/oauth/authorize',
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://auth.calendly.com/oauth/token',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'header',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: '',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: '',
            },
        ];
    }
}
exports.CalendlyOAuth2Api = CalendlyOAuth2Api;
