// n8n-nodes-base/dist/credentials/ConvertApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConvertApi = void 0;
class ConvertApi {
    constructor() {
        this.name = 'convertApi';
        this.displayName = 'ConvertAPI';
        this.documentationUrl = 'convertapi';
        this.icon = 'file:icons/ConvertApi.png';
        this.httpRequestNode = {
            name: 'ConvertAPI',
            docsUrl: 'https://docs.convertapi.com/docs/getting-started',
            apiBaseUrl: 'https://v2.convertapi.com/',
        };
        this.properties = [
            {
                displayName: 'API Token',
                name: 'apiToken',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://v2.convertapi.com',
                url: '/convert/docx/to/pdf',
                ignoreHttpStatusErrors: true,
            },
            rules: [
                {
                    type: 'responseSuccessBody',
                    properties: {
                        key: 'Code',
                        value: 4013,
                        message: 'API Token or Secret is invalid.',
                    },
                },
            ],
        };
    }
}
exports.ConvertApi = ConvertApi;
