// n8n-nodes-base/dist/credentials/DatadogApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatadogApi = void 0;
class DatadogApi {
    constructor() {
        this.name = 'datadogApi';
        this.displayName = 'Datadog API';
        this.documentationUrl = 'datadog';
        this.icon = { light: 'file:icons/Datadog.svg', dark: 'file:icons/Datadog.svg' };
        this.httpRequestNode = {
            name: 'Datadog',
            docsUrl: 'https://docs.datadoghq.com/api/latest/',
            apiBaseUrlPlaceholder: 'https://api.datadoghq.com/api/v1/metrics',
        };
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                required: true,
                type: 'string',
                default: 'https://api.datadoghq.com',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'APP Key',
                name: 'appKey',
                required: false,
                type: 'string',
                default: '',
                typeOptions: { password: true },
                description: 'For some endpoints, you also need an Application key.',
            },
        ];
        this.test = {
            request: {
                baseURL: '={{$credentials.url}}',
                url: '/api/v1/validate',
                method: 'GET',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        requestOptions.headers = {
            'DD-API-KEY': credentials.apiKey,
            'DD-APPLICATION-KEY': credentials.appKey,
        };
        if (!requestOptions.headers['DD-APPLICATION-KEY']) {
            delete requestOptions.headers['DD-APPLICATION-KEY'];
        }
        return requestOptions;
    }
}
exports.DatadogApi = DatadogApi;
