// n8n-nodes-base/dist/credentials/DfirIrisApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DfirIrisApi = void 0;
class DfirIrisApi {
    constructor() {
        this.name = 'dfirIrisApi';
        this.displayName = 'DFIR-IRIS API';
        this.documentationUrl = 'dfiriris';
        this.icon = { light: 'file:icons/DfirIris.svg', dark: 'file:icons/DfirIris.svg' };
        this.httpRequestNode = {
            name: 'DFIR-IRIS',
            docsUrl: 'https://docs.dfir-iris.org/operations/api/',
            apiBaseUrlPlaceholder: 'http://<yourserver_ip>/manage/cases/list',
        };
        this.properties = [
            {
                displayName: 'Base URL',
                name: 'baseUrl',
                type: 'string',
                default: '',
                placeholder: 'e.g. https://localhost',
                description: 'The API endpoints are reachable on the same Address and port as the web interface.',
                required: true,
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'Ignore SSL Issues (Insecure)',
                name: 'skipSslCertificateValidation',
                type: 'boolean',
                default: false,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.baseUrl}}',
                url: '/api/ping',
                method: 'GET',
                skipSslCertificateValidation: '={{$credentials.skipSslCertificateValidation}}',
            },
        };
    }
}
exports.DfirIrisApi = DfirIrisApi;
