// n8n-nodes-base/dist/credentials/DynatraceApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynatraceApi = void 0;
class DynatraceApi {
    constructor() {
        this.name = 'dynatraceApi';
        this.displayName = 'DynatraceAPI';
        this.documentationUrl = 'dynatrace';
        this.icon = { light: 'file:icons/Dynatrace.svg', dark: 'file:icons/Dynatrace.svg' };
        this.httpRequestNode = {
            name: 'Dynatrace',
            docsUrl: 'https://docs.dynatrace.com/docs/dynatrace-api',
            apiBaseUrlPlaceholder: 'https://{your-environment-id}.live.dynatrace.com/api/v2/events',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Api-Token {{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.DynatraceApi = DynatraceApi;
