// n8n-nodes-base/dist/credentials/ElasticSecurityApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticSecurityApi = void 0;
class ElasticSecurityApi {
    constructor() {
        this.name = 'elasticSecurityApi';
        this.displayName = 'Elastic Security API';
        this.documentationUrl = 'elasticSecurity';
        this.properties = [
            {
                displayName: 'Base URL',
                name: 'baseUrl',
                type: 'string',
                default: '',
                placeholder: 'e.g. https://mydeployment.kb.us-central1.gcp.cloud.es.io:9243',
                description: "Referred to as Kibana 'endpoint' in the Elastic deployment dashboard",
                required: true,
            },
            {
                displayName: 'Type',
                name: 'type',
                type: 'options',
                options: [
                    {
                        name: 'API Key',
                        value: 'apiKey',
                    },
                    {
                        name: 'Basic Auth',
                        value: 'basicAuth',
                    },
                ],
                default: 'basicAuth',
            },
            {
                displayName: 'Username',
                name: 'username',
                type: 'string',
                default: '',
                required: true,
                displayOptions: {
                    show: {
                        type: ['basicAuth'],
                    },
                },
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                required: true,
                displayOptions: {
                    show: {
                        type: ['basicAuth'],
                    },
                },
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
                displayOptions: {
                    show: {
                        type: ['apiKey'],
                    },
                },
            },
        ];
        this.test = {
            request: {
                baseURL: '={{$credentials.baseUrl}}',
                url: '/api/endpoint/metadata',
                method: 'GET',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        if (credentials.type === 'apiKey') {
            requestOptions.headers = {
                Authorization: `ApiKey ${credentials.apiKey}`,
            };
        }
        else {
            requestOptions.auth = {
                username: credentials.username,
                password: credentials.password,
            };
            requestOptions.headers = {
                'kbn-xsrf': true,
            };
        }
        return requestOptions;
    }
}
exports.ElasticSecurityApi = ElasticSecurityApi;
