// n8n-nodes-base/dist/credentials/FilescanApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilescanApi = void 0;
class FilescanApi {
    constructor() {
        this.name = 'filescanApi';
        this.displayName = 'Filescan API';
        this.documentationUrl = 'filescan';
        this.icon = { light: 'file:icons/Filescan.svg', dark: 'file:icons/Filescan.svg' };
        this.httpRequestNode = {
            name: 'Filescan',
            docsUrl: 'https://www.filescan.io/api/docs',
            apiBaseUrlPlaceholder: 'https://www.filescan.io/api/system/do-healthcheck',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-Api-Key': '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://www.filescan.io/api',
                url: '/system/do-healthcheck',
                method: 'GET',
            },
        };
    }
}
exports.FilescanApi = FilescanApi;
