// n8n-nodes-base/dist/credentials/IntercomApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntercomApi = void 0;
class IntercomApi {
    constructor() {
        this.name = 'intercomApi';
        this.displayName = 'Intercom API';
        this.documentationUrl = 'intercom';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                    Accept: 'application/json',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.intercom.io',
                url: '/me',
                method: 'GET',
            },
        };
    }
}
exports.IntercomApi = IntercomApi;
