// n8n-nodes-base/dist/credentials/IterableApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IterableApi = void 0;
class IterableApi {
    constructor() {
        this.name = 'iterableApi';
        this.displayName = 'Iterable API';
        this.documentationUrl = 'iterable';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'Region',
                name: 'region',
                type: 'options',
                options: [
                    {
                        name: 'EDC',
                        value: 'https://api.eu.iterable.com',
                    },
                    {
                        name: 'USDC',
                        value: 'https://api.iterable.com',
                    },
                ],
                default: 'https://api.iterable.com',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Api_Key: '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials?.region}}',
                url: '/api/webhooks',
                method: 'GET',
            },
        };
    }
}
exports.IterableApi = IterableApi;
