// n8n-nodes-base/dist/credentials/JiraSoftwareServerPatApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JiraSoftwareServerPatApi = void 0;
class JiraSoftwareServerPatApi {
    constructor() {
        this.name = 'jiraSoftwareServerPatApi';
        this.displayName = 'Jira SW Server (PAT) API';
        this.documentationUrl = 'jira';
        this.properties = [
            {
                displayName: 'Personal Access Token',
                name: 'personalAccessToken',
                typeOptions: { password: true },
                type: 'string',
                default: '',
            },
            {
                displayName: 'Domain',
                name: 'domain',
                type: 'string',
                default: '',
                placeholder: 'https://example.com',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.personalAccessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials?.domain}}',
                url: '/rest/api/2/project',
            },
        };
    }
}
exports.JiraSoftwareServerPatApi = JiraSoftwareServerPatApi;
