// n8n-nodes-base/dist/credentials/MalcoreApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MalcoreApi = void 0;
class MalcoreApi {
    constructor() {
        this.name = 'malcoreApi';
        this.displayName = 'MalcoreAPI';
        this.documentationUrl = 'malcore';
        this.icon = { light: 'file:icons/Malcore.png', dark: 'file:icons/Malcore.png' };
        this.httpRequestNode = {
            name: 'Malcore',
            docsUrl: 'https://malcore.readme.io/reference/upload',
            apiBaseUrlPlaceholder: 'https://api.malcore.io/api/urlcheck',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    apiKey: '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.malcore.io/api',
                url: '/urlcheck',
                method: 'POST',
                body: { url: 'google.com' },
            },
        };
    }
}
exports.MalcoreApi = MalcoreApi;
