// n8n-nodes-base/dist/credentials/MicrosoftSharePointOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MicrosoftSharePointOAuth2Api = void 0;
class MicrosoftSharePointOAuth2Api {
    constructor() {
        this.name = 'microsoftSharePointOAuth2Api';
        this.extends = ['microsoftOAuth2Api'];
        this.icon = {
            light: 'file:icons/SharePoint.svg',
            dark: 'file:icons/SharePoint.svg',
        };
        this.displayName = 'Microsoft SharePoint OAuth2 API';
        this.documentationUrl = 'microsoft';
        this.httpRequestNode = {
            name: 'Microsoft SharePoint',
            docsUrl: 'https://learn.microsoft.com/en-us/sharepoint/dev/apis/sharepoint-rest-graph',
            apiBaseUrlPlaceholder: 'https://{subdomain}.sharepoint.com/_api/v2.0/',
        };
        this.properties = [
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: '=openid offline_access https://{{$self.subdomain}}.sharepoint.com/.default',
            },
            {
                displayName: 'Subdomain',
                name: 'subdomain',
                type: 'string',
                default: '',
                hint: 'You can extract the subdomain from the URL. For example, in the URL "https://tenant123.sharepoint.com", the subdomain is "tenant123".',
            },
        ];
    }
}
exports.MicrosoftSharePointOAuth2Api = MicrosoftSharePointOAuth2Api;
