// n8n-nodes-base/dist/credentials/OpenAiApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAiApi = void 0;
class OpenAiApi {
    constructor() {
        this.name = 'openAiApi';
        this.displayName = 'OpenAi';
        this.documentationUrl = 'openAi';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
            {
                displayName: 'Organization ID (optional)',
                name: 'organizationId',
                type: 'string',
                default: '',
                hint: 'Only required if you belong to multiple organisations',
                description: "For users who belong to multiple organizations, you can set which organization is used for an API request. Usage from these API requests will count against the specified organization's subscription quota.",
            },
            {
                displayName: 'Base URL',
                name: 'url',
                type: 'string',
                default: 'https://api.openai.com/v1',
                description: 'Override the default base URL for the API',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                    'OpenAI-Organization': '={{$credentials.organizationId}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials?.url}}',
                url: '/models',
            },
        };
    }
}
exports.OpenAiApi = OpenAiApi;
