// n8n-nodes-base/dist/credentials/Rapid7InsightVmApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rapid7InsightVmApi = void 0;
class Rapid7InsightVmApi {
    constructor() {
        this.name = 'rapid7InsightVmApi';
        this.displayName = 'Rapid7 InsightVM API';
        this.documentationUrl = 'rapid7insightvm';
        this.icon = {
            light: 'file:icons/Rapid7InsightVm.svg',
            dark: 'file:icons/Rapid7InsightVm.svg',
        };
        this.httpRequestNode = {
            name: 'Rapid7 InsightVM',
            docsUrl: 'https://docs.rapid7.com/',
            apiBaseUrlPlaceholder: 'https://insight.rapid7.com/',
        };
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                required: true,
                type: 'string',
                default: '',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'x-api-key': '={{$credentials.apiKey}}',
                    Accept: 'application/json',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.url}}'.replace(/\/$/, ''),
                url: '/validate',
                method: 'GET',
            },
        };
    }
}
exports.Rapid7InsightVmApi = Rapid7InsightVmApi;
