// n8n-nodes-base/dist/credentials/SysdigApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SysdigApi = void 0;
class SysdigApi {
    constructor() {
        this.name = 'sysdigApi';
        this.displayName = 'Sysdig API';
        this.documentationUrl = 'sysdig';
        this.icon = { light: 'file:icons/Sysdig.Black.svg', dark: 'file:icons/Sysdig.White.svg' };
        this.httpRequestNode = {
            name: 'Sysdig',
            docsUrl: 'https://docs.sysdig.com/en/docs/developer-tools/sysdig-api/',
            apiBaseUrl: 'https://app.us1.sysdig.com',
        };
        this.properties = [
            {
                displayName: 'Access Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.accessToken}}',
                },
            },
        };
    }
}
exports.SysdigApi = SysdigApi;
