// n8n-nodes-base/dist/credentials/TogglApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TogglApi = void 0;
class TogglApi {
    constructor() {
        this.name = 'togglApi';
        this.displayName = 'Toggl API';
        this.documentationUrl = 'toggl';
        this.properties = [
            {
                displayName: 'Email Address',
                name: 'username',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                auth: {
                    username: '={{$credentials.username}}',
                    password: '={{$credentials.password}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.track.toggl.com/api/v9',
                url: '/me',
            },
        };
    }
}
exports.TogglApi = TogglApi;
