// n8n-nodes-base/dist/nodes/Code/Sandbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sandbox = exports.REQUIRED_N8N_ITEM_KEYS = void 0;
exports.getSandboxContext = getSandboxContext;
const events_1 = require("events");
const utils_1 = require("./utils");
const ValidationError_1 = require("./ValidationError");
exports.REQUIRED_N8N_ITEM_KEYS = new Set([
    'json',
    'binary',
    'pairedItem',
    'error',
    'index',
]);
function getSandboxContext(index) {
    const helpers = {
        ...this.helpers,
        httpRequestWithAuthentication: this.helpers.httpRequestWithAuthentication.bind(this),
        requestWithAuthenticationPaginated: this.helpers.requestWithAuthenticationPaginated.bind(this),
    };
    return {
        $getNodeParameter: this.getNodeParameter.bind(this),
        $getWorkflowStaticData: this.getWorkflowStaticData.bind(this),
        helpers,
        ...this.getWorkflowDataProxy(index),
    };
}
class Sandbox extends events_1.EventEmitter {
    constructor(textKeys, helpers) {
        super();
        this.textKeys = textKeys;
        this.helpers = helpers;
    }
    validateRunCodeEachItem(executionResult, itemIndex) {
        if (typeof executionResult !== 'object') {
            throw new ValidationError_1.ValidationError({
                message: `Code doesn't return ${this.getTextKey('object', { includeArticle: true })}`,
                description: `Please return ${this.getTextKey('object', {
                    includeArticle: true,
                })} representing the output item. ('${executionResult}' was returned instead.)`,
                itemIndex,
            });
        }
        if (Array.isArray(executionResult)) {
            const firstSentence = executionResult.length > 0
                ? `An array of ${typeof executionResult[0]}s was returned.`
                : 'An empty array was returned.';
            throw new ValidationError_1.ValidationError({
                message: `Code doesn't return a single ${this.getTextKey('object')}`,
                description: `${firstSentence} If you need to output multiple items, please use the 'Run Once for All Items' mode instead.`,
                itemIndex,
            });
        }
        const [returnData] = this.helpers.normalizeItems([executionResult]);
        this.validateItem(returnData, itemIndex);
        this.validateTopLevelKeys(returnData, itemIndex);
        return returnData;
    }
    validateRunCodeAllItems(executionResult) {
        if (typeof executionResult !== 'object') {
            throw new ValidationError_1.ValidationError({
                message: "Code doesn't return items properly",
                description: `Please return an array of ${this.getTextKey('object', {
                    plural: true,
                })}, one for each item you would like to output.`,
            });
        }
        if (Array.isArray(executionResult)) {
            const mustHaveTopLevelN8nKey = executionResult.some((item) => Object.keys(item).find((key) => exports.REQUIRED_N8N_ITEM_KEYS.has(key)));
            if (mustHaveTopLevelN8nKey) {
                for (let index = 0; index < executionResult.length; index++) {
                    const item = executionResult[index];
                    this.validateTopLevelKeys(item, index);
                }
            }
        }
        const returnData = this.helpers.normalizeItems(executionResult);
        returnData.forEach((item, index) => this.validateItem(item, index));
        return returnData;
    }
    getTextKey(key, options) {
        const response = this.textKeys[key][options?.plural ? 'plural' : 'singular'];
        if (!options?.includeArticle) {
            return response;
        }
        if (['a', 'e', 'i', 'o', 'u'].some((value) => response.startsWith(value))) {
            return `an ${response}`;
        }
        return `a ${response}`;
    }
    validateItem({ json, binary }, itemIndex) {
        if (json === undefined || !(0, utils_1.isObject)(json)) {
            throw new ValidationError_1.ValidationError({
                message: `A 'json' property isn't ${this.getTextKey('object', { includeArticle: true })}`,
                description: `In the returned data, every key named 'json' must point to ${this.getTextKey('object', { includeArticle: true })}.`,
                itemIndex,
            });
        }
        if (binary !== undefined && !(0, utils_1.isObject)(binary)) {
            throw new ValidationError_1.ValidationError({
                message: `A 'binary' property isn't ${this.getTextKey('object', { includeArticle: true })}`,
                description: `In the returned data, every key named 'binary’ must point to ${this.getTextKey('object', { includeArticle: true })}.`,
                itemIndex,
            });
        }
    }
    validateTopLevelKeys(item, itemIndex) {
        Object.keys(item).forEach((key) => {
            if (exports.REQUIRED_N8N_ITEM_KEYS.has(key))
                return;
            throw new ValidationError_1.ValidationError({
                message: `Unknown top-level item key: ${key}`,
                description: 'Access the properties of an item under `.json`, e.g. `item.json`',
                itemIndex,
            });
        });
    }
}
exports.Sandbox = Sandbox;
