// n8n-core/dist/nodes-loader/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.commonCORSParameters = exports.commonPollingParameters = exports.CUSTOM_NODES_CATEGORY = void 0;
const n8n_workflow_1 = require("n8n-workflow");
exports.CUSTOM_NODES_CATEGORY = 'Custom Nodes';
exports.commonPollingParameters = [
    {
        displayName: 'Poll Times',
        name: 'pollTimes',
        type: 'fixedCollection',
        typeOptions: {
            multipleValues: true,
            multipleValueButtonText: 'Add Poll Time',
        },
        default: { item: [{ mode: 'everyMinute' }] },
        description: 'Time at which polling should occur',
        placeholder: 'Add Poll Time',
        options: n8n_workflow_1.cronNodeOptions,
    },
];
exports.commonCORSParameters = [
    {
        displayName: 'Allowed Origins (CORS)',
        name: 'allowedOrigins',
        type: 'string',
        default: '*',
        description: 'Comma-separated list of URLs allowed for cross-origin non-preflight requests. Use * (default) to allow all origins.',
    },
];
