// @n8n/permissions/dist/constants.ee.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RESOURCES = exports.DEFAULT_OPERATIONS = void 0;
exports.DEFAULT_OPERATIONS = ['create', 'read', 'update', 'delete', 'list'];
exports.RESOURCES = {
    annotationTag: [...exports.DEFAULT_OPERATIONS],
    auditLogs: ['manage'],
    banner: ['dismiss'],
    community: ['register'],
    communityPackage: ['install', 'uninstall', 'update', 'list', 'manage'],
    credential: ['share', 'move', ...exports.DEFAULT_OPERATIONS],
    externalSecretsProvider: ['sync', ...exports.DEFAULT_OPERATIONS],
    externalSecret: ['list', 'use'],
    eventBusDestination: ['test', ...exports.DEFAULT_OPERATIONS],
    ldap: ['sync', 'manage'],
    license: ['manage'],
    logStreaming: ['manage'],
    orchestration: ['read', 'list'],
    project: [...exports.DEFAULT_OPERATIONS],
    saml: ['manage'],
    securityAudit: ['generate'],
    sourceControl: ['pull', 'push', 'manage'],
    tag: [...exports.DEFAULT_OPERATIONS],
    user: ['resetPassword', 'changeRole', ...exports.DEFAULT_OPERATIONS],
    variable: [...exports.DEFAULT_OPERATIONS],
    workersView: ['manage'],
    workflow: ['share', 'execute', 'move', ...exports.DEFAULT_OPERATIONS],
    folder: [...exports.DEFAULT_OPERATIONS],
};
