// @n8n/api-types/dist/schemas/insights.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.insightsSummarySchema = exports.insightsSummaryDataSchemas = exports.insightsSummaryUnitSchema = exports.insightsSummaryTypeSchema = void 0;
const zod_1 = require("zod");
exports.insightsSummaryTypeSchema = zod_1.z.enum([
    'total',
    'failed',
    'failureRate',
    'timeSaved',
    'averageRunTime',
]);
exports.insightsSummaryUnitSchema = zod_1.z.enum(['count', 'ratio', 'time']);
exports.insightsSummaryDataSchemas = {
    total: zod_1.z.object({
        value: zod_1.z.number(),
        deviation: zod_1.z.number(),
        unit: zod_1.z.literal('count'),
    }),
    failed: zod_1.z.object({
        value: zod_1.z.number(),
        deviation: zod_1.z.number(),
        unit: zod_1.z.literal('count'),
    }),
    failureRate: zod_1.z.object({
        value: zod_1.z.number(),
        deviation: zod_1.z.number(),
        unit: zod_1.z.literal('ratio'),
    }),
    timeSaved: zod_1.z.object({
        value: zod_1.z.number(),
        deviation: zod_1.z.number(),
        unit: zod_1.z.literal('time'),
    }),
    averageRunTime: zod_1.z.object({
        value: zod_1.z.number(),
        deviation: zod_1.z.number(),
        unit: zod_1.z.literal('time'),
    }),
};
exports.insightsSummarySchema = zod_1.z.object(exports.insightsSummaryDataSchemas).strict();
