// @n8n/n8n-nodes-langchain/dist/credentials/AnthropicApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnthropicApi = void 0;
class AnthropicApi {
    constructor() {
        this.name = 'anthropicApi';
        this.displayName = 'Anthropic';
        this.documentationUrl = 'anthropic';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'x-api-key': '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.anthropic.com',
                url: '/v1/messages',
                method: 'POST',
                headers: {
                    'anthropic-version': '2023-06-01',
                },
                body: {
                    model: 'claude-3-haiku-20240307',
                    messages: [{ role: 'user', content: 'Hey' }],
                    max_tokens: 1,
                },
            },
        };
    }
}
exports.AnthropicApi = AnthropicApi;
