// n8n-nodes-base/dist/nodes/Telegram/tests/Helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMockExecuteFunction = exports.telegramNode = void 0;
const lodash_1 = require("lodash");
exports.telegramNode = {
    id: 'b3039263-29ad-4476-9894-51dfcc5a706d',
    name: 'Telegram node',
    typeVersion: 1.2,
    type: 'n8n-nodes-base.telegram',
    position: [0, 0],
    parameters: {
        resource: 'callback',
        operation: 'answerQuery',
    },
};
const createMockExecuteFunction = (nodeParameters) => {
    const fakeExecuteFunction = {
        getInputData() {
            return [{ json: {} }];
        },
        getNodeParameter(parameterName, _itemIndex, fallbackValue, options) {
            const parameter = options?.extractValue ? `${parameterName}.value` : parameterName;
            return (0, lodash_1.get)(nodeParameters, parameter, fallbackValue);
        },
        getNode() {
            return exports.telegramNode;
        },
        helpers: {},
        continueOnFail: () => false,
    };
    return fakeExecuteFunction;
};
exports.createMockExecuteFunction = createMockExecuteFunction;
