// n8n-nodes-base/dist/nodes/Form/formCompletionUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderFormCompletion = void 0;
const utils_1 = require("./utils");
const renderFormCompletion = async (context, res, trigger) => {
    const completionTitle = context.getNodeParameter('completionTitle', '');
    const completionMessage = context.getNodeParameter('completionMessage', '');
    const redirectUrl = context.getNodeParameter('redirectUrl', '');
    const options = context.getNodeParameter('options', {});
    const responseText = context.getNodeParameter('responseText', '');
    let title = options.formTitle;
    if (!title) {
        title = context.evaluateExpression(`{{ $('${trigger?.name}').params.formTitle }}`);
    }
    const appendAttribution = context.evaluateExpression(`{{ $('${trigger?.name}').params.options?.appendAttribution === false ? false : true }}`);
    res.render('form-trigger-completion', {
        title: completionTitle,
        message: completionMessage,
        formTitle: title,
        appendAttribution,
        responseText: (0, utils_1.sanitizeHtml)(responseText),
        dangerousCustomCss: (0, utils_1.sanitizeCustomCss)(options.customCss),
        redirectUrl,
    });
    return { noWebhookResponse: true };
};
exports.renderFormCompletion = renderFormCompletion;
