// @n8n/n8n-nodes-langchain/dist/nodes/vector_store/VectorStoreInMemory/VectorStoreInMemory.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VectorStoreInMemory = void 0;
const createVectorStoreNode_1 = require("../shared/createVectorStoreNode/createVectorStoreNode");
const MemoryVectorStoreManager_1 = require("../shared/MemoryVectorStoreManager");
const insertFields = [
    {
        displayName: 'The embedded data are stored in the server memory, so they will be lost when the server is restarted. Additionally, if the amount of data is too large, it may cause the server to crash due to insufficient memory.',
        name: 'notice',
        type: 'notice',
        default: '',
    },
    {
        displayName: 'Clear Store',
        name: 'clearStore',
        type: 'boolean',
        default: false,
        description: 'Whether to clear the store before inserting new data',
    },
];
class VectorStoreInMemory extends (0, createVectorStoreNode_1.createVectorStoreNode)({
    meta: {
        displayName: 'Simple Vector Store',
        name: 'vectorStoreInMemory',
        description: "Work with your data in a Simple Vector Store. Don't use this for production usage.",
        icon: 'fa:database',
        iconColor: 'black',
        docsUrl: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/root-nodes/n8n-nodes-langchain.vectorstoreinmemory/',
    },
    sharedFields: [
        {
            displayName: 'Memory Key',
            name: 'memoryKey',
            type: 'string',
            default: 'vector_store_key',
            description: 'The key to use to store the vector memory in the workflow data. The key will be prefixed with the workflow ID to avoid collisions.',
        },
    ],
    insertFields,
    loadFields: [],
    retrieveFields: [],
    async getVectorStoreClient(context, _filter, embeddings, itemIndex) {
        const workflowId = context.getWorkflow().id;
        const memoryKey = context.getNodeParameter('memoryKey', itemIndex);
        const vectorStoreSingleton = MemoryVectorStoreManager_1.MemoryVectorStoreManager.getInstance(embeddings);
        return await vectorStoreSingleton.getVectorStore(`${workflowId}__${memoryKey}`);
    },
    async populateVectorStore(context, embeddings, documents, itemIndex) {
        const memoryKey = context.getNodeParameter('memoryKey', itemIndex);
        const clearStore = context.getNodeParameter('clearStore', itemIndex);
        const workflowId = context.getWorkflow().id;
        const vectorStoreInstance = MemoryVectorStoreManager_1.MemoryVectorStoreManager.getInstance(embeddings);
        await vectorStoreInstance.addDocuments(`${workflowId}__${memoryKey}`, documents, clearStore);
    },
}) {
}
exports.VectorStoreInMemory = VectorStoreInMemory;
